filebeat/inputsource/common/dgram/server.go

```diff
@@ -0,0 +1,194 @@
+
+type ConnectionHandler func(context.Context, ReadFromConnection) error
+
+type ListenerFactory func() (ReadFromConnection, error)
+
+// MetadataFunc defines callback executed when a line is read from the split handler.
+type MetadataFunc func(net.Conn) inputsource.NetworkMetadata
+
+type ListenerConfig struct {
+	Timeout        time.Duration
+	MaxMessageSize cfgtype.ByteSize
+}
+
+type Listener struct {
+	log      *logp.Logger
+	config   *ListenerConfig
+	listener ListenerFactory
+	connect  HandlerFactory
+	tg       unison.TaskGroup
+}
+
+func NewListener(
+	f inputsource.Family,
+	path string,
+	connect HandlerFactory,
+	listenerFactory ListenerFactory,
+	config *ListenerConfig,
+) *Listener {
+	return &Listener{
+		log:      logp.NewLogger(f.String()),
+		config:   config,
+		listener: listenerFactory,
+		connect:  connect,
+		tg:       unison.TaskGroup{},
+	}
+}
+
+// DatagramReaderFactory allows creation of a handler that has splitting capabilities.
+func DatagramReaderFactory(
+	family inputsource.Family,
+	logger *logp.Logger,
+	callback inputsource.NetworkFunc,
+) HandlerFactory {
+	return func(config ListenerConfig) ConnectionHandler {
+		return ConnectionHandler(func(ctx context.Context, conn ReadFromConnection) error {
+			//metadata := metadataCallback(conn)
+			for ctx.Err() == nil {
+
+				buffer := make([]byte, config.MaxMessageSize)
+				conn.SetDeadline(time.Now().Add(config.Timeout))
+
+				// If you are using Windows and you are using a fixed buffer and you get a datagram which
+				// is bigger than the specified size of the buffer, it will return an `err` and the buffer will
+				// contains a subset of the data.
+				//
+				// On Unix based system, the buffer will be truncated but no error will be returned.
+				length, addr, err := conn.ReadFrom(buffer)
+				if err != nil {
+					// don't log any deadline events.
+					e, ok := err.(net.Error)
+					if ok && e.Timeout() {
+						continue
+					}
+
+					// Closed network error string will never change in Go 1.X
+					// https://github.com/golang/go/issues/4373
+					opErr, ok := err.(*net.OpError)
+					if ok && strings.Contains(opErr.Err.Error(), "use of closed network connection") {
+						logger.Info("Connection has been closed")
+						return nil
+					}
+
+					logger.Errorf("Error reading from the socket %s", err)
+
+					// On Windows send the current buffer and mark it as truncated.
+					// The buffer will have content but length will return 0, addr will be nil.
+					if isLargerThanBuffer(err) {
+						callback(buffer, inputsource.NetworkMetadata{RemoteAddr: addr, Truncated: true})
+						continue
+					}
+				}
+
+				if length > 0 {
+					callback(buffer[:length], inputsource.NetworkMetadata{RemoteAddr: addr})
+				}
+			}
+			return nil
+		})
+	}
+}
+
+func isLargerThanBuffer(err error) bool {
+	if runtime.GOOS != "windows" {
+		return false
+	}
+	return strings.Contains(err.Error(), windowErrBuffer)
+}
+
+func (l *Listener) Run(ctx context.Context) error {
```

Comment:
If Run returns on error the input stops hard. Are there some errors we can handle gracefully? e.g.
- unix file can not be opened by current user => retry with backoff (user can adjust permissions)
- another filebeat instance is listening on unix socket => retry with backoff (other FB instance might go down => HA setup?)
- UDP port is currently take. Similar to unix socket, if it is taken by another filebeat instance we might want to retry.
- Do want to allow the handler to take down the server? e.g. distinguish between critical or non-critical errors? Or do we mandate that the handler MUST ONLY return an error if it is not Recoverable from within the handler (IO error). On IO error, can we recover by trying to reopen the connection and listen?

All in all, the knowledge if an error is 'fatal' for the input or not lives in the actual inputsource implementation, not here. It looks like you just moved the logic from the current UDP server (refactoring). Maybe let us create an issue instead of fixing/improving error handling here. I would presume that we can improve error handling in the TCP/unix-streaming inputs as well.